cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+	}
+
+	selectedPods := make([]*apiv1.Pod, len(data.podList), len(data.podList))
+	// i is index in *output* list.
+	i := 0
+	for _, pod := range data.podList {
+		if podFilter(pod) && selector.Matches(labels.Set(pod.Labels)) {
+			selectedPods[i] = pod
+			i++
+		}
+	}
+	selectedPods = selectedPods[:i]
+	return selectedPods, nil
+}
+
+func (data *internalDeltaSnapshotData) Pods() schedulerlisters.PodLister {
+	return (*internalDeltaSnapshotDataPodLister)(data)
+}
+
+func (data *internalDeltaSnapshotData) NodeInfos() schedulerlisters.NodeInfoLister {
+	return (*internalDeltaSnapshotDataNodeLister)(data)
+}
+
+// NewEmptySnapshot initializes a Snapshot struct and returns it.
+func newInternalDeltaSnapshotData() *internalDeltaSnapshotData {
+	return &internalDeltaSnapshotData{
+		nodeInfoMap:      make(map[string]*schedulernodeinfo.NodeInfo),
+		deletedNodeInfos: make(map[string]bool),
+	}
+}
+
+// Convenience method to avoid writing loop for adding nodes.
+func (data *internalDeltaSnapshotData) addNodes(nodes []*apiv1.Node) error {
+	for _, node := range nodes {
+		if err := data.addNode(node); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addNode(node *apiv1.Node) error {
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	if err := nodeInfo.SetNode(node); err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo: %v", err)
+	}
+	return data.addNodeInfo(nodeInfo)
+}
+
+func (data *internalDeltaSnapshotData) addNodeInfo(nodeInfo *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.nodeInfoMap[nodeInfo.Node().Name]; found {
+		return fmt.Errorf("node %s already in snapshot", nodeInfo.Node().Name)
+	}
+	data.nodeInfoMap[nodeInfo.Node().Name] = nodeInfo
+	if data.nodeInfoList != nil {
+		data.nodeInfoList = append(data.nodeInfoList, nodeInfo)
+	}
+	if data.completeNodeInfoList != nil {
+		data.completeNodeInfoList = append(data.completeNodeInfoList, nodeInfo)
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) clearCaches() {
+	data.nodeInfoList = nil
+	data.completeNodeInfoList = nil
+	data.clearPodCaches()
+}
+
+func (data *internalDeltaSnapshotData) clearPodCaches() {
+	data.podList = nil
+	data.rawPods = nil
+	data.rawPodsTotal = 0
+	data.havePodsWithAffinity = nil
+}
+
+func (data *internalDeltaSnapshotData) updateNode(node *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.nodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+
+	return data.addNodeInfo(node)
+}
+
+func (data *internalDeltaSnapshotData) removeNode(nodeName string) error {
+	_, found := data.nodeInfoMap[nodeName]
+	if found {
+		// If node was added or modified within this delta, delete this change.
+		delete(data.nodeInfoMap, nodeName)
+	}
+
+	ni, err := data.baseData.GetNodeInfo(nodeName)
+	if err == nil && ni != nil {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if err != nil && !found {
```

Comment:
I do not like the fact that we silently assume that `err != nil` means `not found`.
Maybe we should use some internal version of `GetNodeInfo` which explicitly return if node was present or not. Not treating fact that it is not present as an error?

Reply:
Changed to return (*schedulernodeinfo.NodeInfo, bool).

Another possibility would be to return just *schedulernodeinfo.NodeInfo, and assume nil means nothing. WDYT?